src/serial.rs

```diff
@@ -197,18 +328,43 @@ where
         Tx: TxPin<Usart>,
         Rx: RxPin<Usart>,
     {
+        use self::config::*;
+
+        let config = config.into();
+
+        // Enable USART peripheral for any further interaction.
         Usart::enable_clock(apb);
+        // Disable USART because some configuration bits could only be written
+        // in this state.
+        usart.cr1.modify(|_, w| w.ue().disabled());
 
-        let brr = Usart::clock(&clocks).integer() / baud_rate.integer();
+        let brr = Usart::clock(&clocks).integer() / config.baudrate.integer();
         crate::assert!(brr >= 16, "impossible baud rate");
         usart.brr.write(|w| w.brr().bits(brr as u16));
 
+        let m0 = match config.wordlength {
+            WordLength::DataAndParityBits8 => M_A::BIT8,
+            WordLength::DataAndParityBits9 => M_A::BIT9,
+        };
```

Comment:
Oh I see, this is not possible, because WordLength config is split up to M0 and M1, correct? Is impl `Into<(M0_A, M1_A)> for WordLength` possible? :D

Reply:
I thought about it a bit:

My proposal would be to add this function, which changes the wordlength depending on the parity chosen:

```rust
impl Config {
    pub fn parity(self, parity: Parity) -> Self {
        self.parity = match parity {
            Parity::None => parity,
            Parity::Even | Parity::Odd => {
                self.wordlength = Worlength::DataAndParityBits9
                parity
            }
        }
    }
}
```

With this change, the rest should be thought out again. The `parity_none`, `parity_even` and `parity_odd` functions could be removed, and I think we should maybe remove or restrict pub access to the fields off the Config struct? This would need getter and setter functions though, or more complicated patterns as the build pattern or similar. I'm not sure, if this is the right way to go. Maybe a comment to steer away the user from directly changing the fields is sufficient?

Reply:
> Oh I see, this is not possible, because WordLength config is split up to M0 and M1, correct? Is impl `Into<(M0_A, M1_A)> for WordLength` possible? :D

I will give it a try.

Reply:
> My proposal would be to add this function, which changes the wordlength depending on the parity chosen:

I like your idea! I was thinking about this for some time but did not came up with a satisfying solution. Coming up with a sound configuration seems a challenge to me as not all logical configurations map to feasible ones. For example selecting 9 data bits plus a parity bit is not possible.

Checking for a safe configuration could be done in `Config` and before instantiating `Serial`. Just dreaming: Do you think it is possible to get a configuration with compile-time constant values checked at compile time? In this case all input parameters to the config could be logical values (for example 8 data bits and one party bit) which will be boiled down to a physical configuration (9 data and parity bits, parity mode and PCE enabled).

And for runtime configuration there could be a variant returning an `Option`/`Result`.

The more I'm writing here the more I'm thinking that `Config` should be a pure builder to sort out unsound configurations.

Let's work something out here. When talking about a serial abstraction, I would prefer not needing to have a deep dive into the USART specialties for getting to a working configuration.

Reply:
And nothing prevents us from providing an unsafe configuration option which allows to specify all the physical parameters directly to cover situations we have not thought about and might be needed for _honey, I know what I'm doing_.

Reply:
> For example selecting 9 data bits plus a parity bit is not possible.

And also selecting 9 data bits without parity is also not possible right now, because the serial implementation assumes 8 bit Word width, so all function do use `u8`. To support 9 bit data, we need another implementation with `u16` which would fit `u8`. This is what the stm32f4xx-hal is doing IIRC. 

But I would suggest for now to keep it mind to keep this possible open for the future, but you do not have to implement it now. All in all I like a iterative approach better, which come with some more breaking changes but will result into a ergonic design instead of trying to catch all use cases at once. 

> Checking for a safe configuration could be done in `Config` and before instantiating `Serial`. Just dreaming: Do you think it is possible to get a configuration with compile-time constant values checked at compile time?

These are ideas sound really similar to things I have in mind for the rcc clock configuration. My idea is to leverage the builder pattern, where the building of the function could all be `const fn`'s so that these could be calculated at compile time. But this comes with some ergonomic overhead and I think `const fn`'s are not far enough in rustc to really be usable for this case. 

> 
> And for runtime configuration there could be a variant returning an `Option`/`Result`.
> 
> The more I'm writing here the more I'm thinking that `Config` should be a pure builder to sort out unsound configurations.

Yeah this is definitely the most robust way. 

> 
> Let's work something out here. When talking about a serial abstraction, I would prefer not needing to have a deep dive into the USART specialties for getting to a working configuration.

I suggest to start small. As it is not possible right now to support 9 bit Word width and 7 bits we can ignore right now, what about removing the word-length setting from the Config struct for now and just make parity configurable. Then we change the wordLength according to the parity, so that no unexpected things happen (like reducing the word width to 7 bits, when enabling parity). What do you think?

Reply:
> Oh I see, this is not possible, because WordLength config is split up to M0 and M1, correct? Is impl `Into<(M0_A, M1_A)> for WordLength` possible? :D

I gave it a try (with different types) and it seems possible:
``` rust
impl Into<(M_A, PS_A, PCE_A)> for Parity {
    fn into(self) -> (M_A, PS_A, PCE_A) {
         match self {
            Parity::ParityNone => (M_A::BIT8, PS_A::EVEN, PCE_A::DISABLED),
            Parity::ParityEven => (M_A::BIT9, PS_A::EVEN, PCE_A::ENABLED),
            Parity::ParityOdd => (M_A::BIT9, PS_A::ODD, PCE_A::ENABLED),
        }
    }
}
```
``` rust
let (m0, ps, pce): (M_A, PS_A, PCE_A) = config.parity.into();
```
But there are two things my gut instinct opposes against:

1. If `Serial` one day exposes the USART internals, wouldn't this be the right point in time for exposing such types from its API and not before?
2. For the example with `Parity`, the tuple type would look pretty arbitrary to me as a regular user. Would this type make sens to you?

As there is currently little code for such conversions in `Serial::new`, I would keep it there until the types are otherwise exposed/used by `Serial`'s public API.

Reply:
> I suggest to start small. As it is not possible right now to support 9 bit Word width and 7 bits we can ignore right now, what about removing the word-length setting from the Config struct for now and just make parity configurable. Then we change the wordLength according to the parity, so that no unexpected things happen (like reducing the word width to 7 bits, when enabling parity). What do you think?

This looks like a pretty appealing way to me. I've refactored my changes in that sense.

Reply:
> As there is currently little code for such conversions in Serial::new, I would keep it there until the types are otherwise exposed/used by Serial's public API.

Yeah, good point. As it is only used in one place and the `Into` trait implementation would make it public, let's not use it. And the ergonomic "improvements" are not really improvements. Thanks for the experiment.